src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/1.sql

```diff
@@ -273,6 +273,392 @@ ON dbo.CompartmentAssignment
+    ReferenceResourceId varchar(64) NOT NULL,
+    ReferenceResourceVersion int NULL,
+    IsHistory bit NOT NULL,
+) WITH (DATA_COMPRESSION = PAGE)
+
+CREATE CLUSTERED INDEX IXC_ReferenceSearchParam
+ON dbo.ReferenceSearchParam
+(
+    ResourceSurrogateId,
+    SearchParamId,
+    ReferenceResourceTypeId,
+    ReferenceResourceId,
+    BaseUri,
+    ReferenceResourceVersion
+)
+
+CREATE NONCLUSTERED INDEX IX_ReferenceSearchParam_SearchParamId_ReferenceResourceTypeId_ReferenceResourceId_BaseUri_ReferenceResourceVersion
+ON dbo.ReferenceSearchParam
+(
+    SearchParamId,
+    ReferenceResourceTypeId,
+    ReferenceResourceId,
+    BaseUri,
+    ReferenceResourceVersion
+) 
+WHERE IsHistory = 0
+WITH (DATA_COMPRESSION = PAGE)
+
+GO
+
+/*************************************************************
+    Token Search Param
+**************************************************************/
+
+CREATE TYPE dbo.TokenSearchParamTableType_1 AS TABLE  
+(
+    SearchParamId smallint NOT NULL,
+    SystemId int NULL,
+    Code varchar(128) NOT NULL
+)
+
+CREATE TABLE dbo.TokenSearchParam
+(
+    ResourceSurrogateId bigint NOT NULL,
+    SearchParamId smallint NOT NULL,
+    SystemId int NULL,
+    Code varchar(128) NOT NULL,
+    IsHistory bit NOT NULL,
+) WITH (DATA_COMPRESSION = PAGE)
+
+CREATE CLUSTERED INDEX IXC_TokenSearchParam
+ON dbo.TokenSearchParam
+(
+    ResourceSurrogateId,
+    SearchParamId,
+    Code,
+    SystemId
+)
+
+CREATE NONCLUSTERED INDEX IX_TokenSeachParam_SearchParamId_Code_SystemId 
+ON dbo.TokenSearchParam
+(
+    SearchParamId,
+    Code,
+    SystemId
+) 
+WHERE IsHistory = 0
+WITH (DATA_COMPRESSION = PAGE)
+
+GO
+
+/*************************************************************
+    Token Text
+**************************************************************/
+
+CREATE TYPE dbo.TokenTextTableType_1 AS TABLE  
+(
+    SearchParamId smallint NOT NULL,
+    Text nvarchar(400) COLLATE Latin1_General_CI_AI NOT NULL
+)
+
+CREATE TABLE dbo.TokenText
+(
+    ResourceSurrogateId bigint NOT NULL,
+    SearchParamId smallint NOT NULL,
+    Text nvarchar(400) COLLATE Latin1_General_CI_AI NOT NULL,
+    IsHistory bit NOT NULL
+) WITH (DATA_COMPRESSION = PAGE)
+
+CREATE CLUSTERED INDEX IXC_TokenText
+ON dbo.TokenText
+(
+    ResourceSurrogateId,
+    SearchParamId,
+    Text
+)
+
+CREATE NONCLUSTERED INDEX IX_TokenText_SearchParamId_Text
+ON dbo.TokenText
+(
+    SearchParamId,
+    Text
+) 
+WHERE IsHistory = 0
+WITH (DATA_COMPRESSION = PAGE)
+
+GO
+
+/*************************************************************
+    String Search Param
+**************************************************************/
+
+CREATE TYPE dbo.StringSearchParamTableType_1 AS TABLE  
+(
+    SearchParamId smallint NOT NULL,
+    Text nvarchar(400) COLLATE Latin1_General_CI_AI NOT NULL
+)
+
+CREATE TABLE dbo.StringSearchParam
+(
+    ResourceSurrogateId bigint NOT NULL,
+    SearchParamId smallint NOT NULL,
+    Text nvarchar(400) COLLATE Latin1_General_CI_AI NOT NULL,
```

Comment:
How did we pick 400?  What happens if we attempt to store a longer value?

Reply:
Great question. This is not set in stone yet. Clustered and nonclustered cannot exceed 900 and 1700 bytes respectively. Including the resource surrogate ID in the clustered index probably means that there is not much to be gained by including the Text column. So we can make it longer if we only put it in nonclustered indexes.
But there still will be a limit. To lift the limit, we could have an overflow column that is not indexed that we only reach into when predicates are exceed the key column length.
Right now, the resource insert will fail if there is a string search param value > 400. 
FWIW, Hapi has a max length of 200.